src/frontend/persistent-connection-pair.js

```diff
@@ -0,0 +1,348 @@
+	},
+	
+	_onReconnected: function () {
+		logger.log(this._options.logPrefix + 'Reconnected.');
+		this.emit('reconnected');
+	},
+	
+	_onData: function (data) {
+		//logger.log(this._options.logPrefix + 'Data:', data);
+		this.emit('data', data);
+	},
+	
+	_onEnd: function () {
+		logger.warn(this._options.logPrefix + 'End.');
+		this.emit('end');
+	},
+	
+	_onError: function (error) {
+		logger.error(this._options.logPrefix + 'Error:', error);
+		this.emit('error', error);
+	},
+	
+	_onConnConnecting: function (descriptor) {
+		var _this = this;
+		if (descriptor.index === _this._frontIndex) {
+			// This can only happen on front connection reconnect.
+		}
+		else {
+			if (!_this._isConnected) {
+				_this._isConnecting = true;
+				
+				logger.log(_this._options.logPrefix + 'Connecting: ' +
+					'front = [' + (_this._frontIndex + 1) + '], back = [' + (_this._backIndex + 1) + '].');
+				
+				_this._onConnecting();
+			}
+			else {
+				_this._isReconnecting = true;
+				
+				logger.log(_this._options.logPrefix + 'Reconnecting: ' +
+					'front = [' + (_this._frontIndex + 1) + '], back = [' + (_this._backIndex + 1) + '].');
+				
+				_this._onReconnecting();
+			}
+		}
+	},
+	
+	_onConnConnected: function (descriptor) {
+		var _this = this;
+		if (descriptor.index === _this._frontIndex) {
+			// This can only happen on front connection reconnect.
+		}
+		else {
+			var frontIndex = _this._frontIndex;
+			var wasConnected = _this._isConnected;
+			
+			// Swap the connections.
+			logger.log(_this._options.logPrefix + 'Swap connections: ' +
+				'front = [' + (descriptor.index + 1) + '], back = [' + (frontIndex + 1) + '].');
+			
+			_this._backIndex = frontIndex;
+			_this._frontIndex = descriptor.index;
+			
+			_this._isReconnecting = false;
+			
+			if (wasConnected) {
+				// Disconnect the front connection.
+				logger.log(_this._options.logPrefix + 'Disconnect: ' +
+					'front = [' + (frontIndex + 1) + '].');
+				
+				var frontDescriptor = _this._connections[frontIndex];
+				frontDescriptor.disabled = true;
+				frontDescriptor.conn.disconnect();
+				frontDescriptor.disabled = false;
+				
+				_this._onReconnected();
+			}
+			else {
+				_this._isConnecting = false;
+				_this._isConnected = true;
+				_this._onConnected();
+			}
+		}
+	},
+	
+	_onConnData: function (descriptor, data) {
+		var _this = this;
+		if (descriptor.index === _this._frontIndex) {
+			_this._onData(data);
+		}
+		else {
+			// This should not happen.
+			throw new Error('PersistentConnectionPair#_onConnData: Got data from back connection.');
+		}
+	},
+	
+	_onConnError: function (descriptor, error) {
+		var _this = this;
+		if (descriptor.index === _this._frontIndex) {
+			_this._dropConnection(error);
+		}
+		else {
+			if (_this._isConnecting || _this._isReconnecting) {
+				_this._dropConnection(error);
+			}
+			else if (_this._isConnected) {
+				descriptor.conn.reconnect();
+			}
+		}
+	},
+	
+	_onConnEnd: function (descriptor) {
```

Comment:
They were quite different while debugging until I finally came to this code. Fixed.